lib/Conversion/ImportVerilog/Expressions.cpp

```diff
@@ -1024,55 +1024,127 @@ struct RvalueExprVisitor : public ExprVisitor {
-      if (!value)
-        return {};
-      elements.push_back(value);
-    }
-    for (unsigned replIdx = 1; replIdx < replCount; ++replIdx)
-      for (unsigned elementIdx = 0; elementIdx < elementCount; ++elementIdx)
-        elements.push_back(elements[elementIdx]);
+    const auto &elts = expr.elements();
 
     // Handle integers.
     if (auto intType = dyn_cast<moore::IntType>(type)) {
-      assert(intType.getWidth() == elements.size());
-      std::reverse(elements.begin(), elements.end());
-      return moore::ConcatOp::create(builder, loc, intType, elements);
+      auto elements = convertElements(expr, {}, replCount);
+
+      if (failed(elements))
+        return {};
+
+      assert(intType.getWidth() == elements->size());
+      std::reverse(elements->begin(), elements->end());
+      return moore::ConcatOp::create(builder, loc, intType, *elements);
     }
 
     // Handle packed structs.
     if (auto structType = dyn_cast<moore::StructType>(type)) {
-      assert(structType.getMembers().size() == elements.size());
-      return moore::StructCreateOp::create(builder, loc, structType, elements);
+      SmallVector<Type> expectedTy;
+      expectedTy.reserve(structType.getMembers().size());
+      for (auto member : structType.getMembers())
+        expectedTy.push_back(member.type);
+
+      FailureOr<SmallVector<Value>> elements;
+      if (expectedTy.size() == elts.size())
+        elements = convertElements(expr, expectedTy, replCount);
+      else
+        elements = convertElements(expr, {}, replCount);
+
+      if (failed(elements))
+        return {};
+
+      assert(structType.getMembers().size() == elements->size());
+      return moore::StructCreateOp::create(builder, loc, structType, *elements);
     }
 
     // Handle unpacked structs.
     if (auto structType = dyn_cast<moore::UnpackedStructType>(type)) {
-      assert(structType.getMembers().size() == elements.size());
-      return moore::StructCreateOp::create(builder, loc, structType, elements);
+      SmallVector<Type> expectedTy;
+      expectedTy.reserve(structType.getMembers().size());
+      for (auto member : structType.getMembers())
+        expectedTy.push_back(member.type);
+
+      FailureOr<SmallVector<Value>> elements;
+      if (expectedTy.size() == elts.size())
+        elements = convertElements(expr, expectedTy, replCount);
+      else
+        elements = convertElements(expr, {}, replCount);
+
+      if (failed(elements))
+        return {};
+
+      assert(structType.getMembers().size() == elements->size());
+
+      return moore::StructCreateOp::create(builder, loc, structType, *elements);
     }
 
     // Handle packed arrays.
     if (auto arrayType = dyn_cast<moore::ArrayType>(type)) {
-      assert(arrayType.getSize() == elements.size());
-      return moore::ArrayCreateOp::create(builder, loc, arrayType, elements);
+      auto elements = convertElements(expr, {}, replCount);
```

Comment:
Would it make sense to pass the expected element type in for arrays? They might suffer from the same issue as the struct patterns. You could have an `std::variant<Type, ArrayRef<Type>>` on `convertElements` to distinguish between a uniform type across all elements, and a specific type per element. Or maybe a separate `Type uniformType` operand? Or treat `expectedTypes` of size 1 as the uniform type for all elements?

Reply:
I can definitely adapt that. I haven't been able to produce the same bug for Array conversions though - should we still go for it?

Reply:
Refactored to use a `std::variant<Type, ArrayRef<Type>>` and applying it to array elements as well. 
Since int-casts are anyway on the "direct and happy path" in `convertRvalueExpression` and the current code treats `IntType` assignment over optionally multiple RValues, I left those out as it seems unsafe (e.g. you could concat a 2-bit and two 1-bit expressions to a 4 -bit expression, explicit casts would mess this up, probably).